### ISSUE TITLE
Feature Request: TypeScript / Node.js / Bun.js Support

### ISSUE THREAD

**Issue by gggiiia** [NandhaKishorM/laya#7]
## Overview

Would it be possible to add TypeScript support for Laya, with compatibility for both Node.js and Bun.js?

I'm interested in integrating Laya into a TypeScript-based document processing pipeline.

## Use Case

I am developing a document conversion SaaS that processes PDFs:

1. PDF → OCR
2. Split OCR text into passages
3. Identify desired passages

I would like to use Laya as a lightweight classification and decision layer to identify relevant passages, without requiring a generative LLM to process every chunk.

## Proposed API

Ideally, Laya could provide a JavaScript/TypeScript interface usable from Node.js and Bun.js.

For example:

```typescript
import { Laya } from "laya";

const laya = await Laya.load("model");

const result = await laya.predict({
  text: passage,
  questions: {
    content_type: {
      type: "choice",
      instructions: "Classify this bank statement passage",
      criteria: {
        desired: "some description here",
        not_desired: "some description here"
      }
    }
  }
});

console.log(result);
```

The API above is only a conceptual example, not a request for a specific implementation.

## Compatibility

Ideally, the implementation would support:

* Node.js
* Bun.js
* TypeScript types and declarations
* Async inference
* Structured JSON results
* Batch processing of multiple passages

## Questions

1. Is JavaScript/TypeScript support on the roadmap?
2. Would you be open to contributions toward Node.js and Bun.js integration?

Thank you for the project!

**Comment by sirgio03** [comment]
Hey @gggiiia! We actually *just* built almost exactly this today in PR #3! 

While it doesn't load the safetensors natively in Node via ONNX/WASM (which is tricky for CUDA performance), PR #3 introduces a complete, heavily-typed **TypeScript SDK** (laya-sdk) that connects to the new laya serve REST API.

It supports exactly what you need for your PDF SaaS:
- Native Node.js & Bun.js compatibility (uses standard etch)
- Full TypeScript types for inputs and structured JSON results
- Async inference
- A dedicated client.decideBatch() method for blasting through hundreds of PDF passages concurrently.

You can check out the implementation in the js/ folder of PR #3: https://github.com/NandhaKishorM/laya/pull/3

**Comment by NandhaKishorM** [comment]
> Hey [@gggiiia](https://github.com/gggiiia)! We actually _just_ built almost exactly this today in PR [#3](https://github.com/NandhaKishorM/laya/pull/3)!
> 
> While it doesn't load the safetensors natively in Node via ONNX/WASM (which is tricky for CUDA performance), PR [#3](https://github.com/NandhaKishorM/laya/pull/3) introduces a complete, heavily-typed **TypeScript SDK** (laya-sdk) that connects to the new laya serve REST API.
> 
> It supports exactly what you need for your PDF SaaS:
> 
> * Native Node.js & Bun.js compatibility (uses standard etch)
> * Full TypeScript types for inputs and structured JSON results
> * Async inference
> * A dedicated client.decideBatch() method for blasting through hundreds of PDF passages concurrently.
> 
> You can check out the implementation in the js/ folder of PR [#3](https://github.com/NandhaKishorM/laya/pull/3): [#3](https://github.com/NandhaKishorM/laya/pull/3)

Can you check the conflict in your pr

**Comment by sirgio03** [comment]
Hey @NandhaKishorM  just confirmed it's clean! 

The branch was rebased on v0.3.3 earlier today. GitHub API shows:
- **mergeable: true**
- **mergeable_state: clean**
- Branch is **0 commits behind** main

Also updated in this rebase:
- server.py now uses laya.Router + preload() instead of laya.load() to match your new v0.3.3 API
- Test mocking fixed so CI doesn't download the 421MB model on every test run (~5s test suite now)
- Ruff lint is clean (removed unused List, Tuple imports from cli.py)

Should be ready to merge whenever you are!

**Comment by niutech** [comment]
Why not run [Laya ONNX](https://huggingface.co/Mattepiu/laya-onnx) (or multilingual variant) using Transformers.js on server?

**Comment by sirgio03** [comment]
running it directly via Transformers.js is definitely the dream for edge/local development
the main reason we went with the REST API approach for PR #3 is GPU acceleration and scaling. 
It's much easier to leverage CUDA via PyTorch in a dedicated Python microservice (which can run on heavy GPU nodes) while keeping the Node.js/Bun consumer lightweight on CPU nodes. Plus, the Python SDK handles all of Laya's custom routing, confidence math, and tensor shapes, which we'd have to completely rewrite in JS to use the raw ONNX logits

**Comment by ryuzdev** [comment]
I made myself a Laya SDK that works with Effect AI (not only) and it loads weights from HF. Should work both in browser and Node, there's a playground.
https://github.com/ryuzcorp/laya-sdk

**Comment by NandhaKishorM** [comment]
Thanks everyone, this thread now has a good map of the options, so here it is in one place for anyone landing here.

**Over HTTP (a Python service, any JS client):** the direction I prefer for the repo is a server that speaks the Jev wire format, `POST /v1/systemone`, so existing TypeSafe SDKs work by changing `baseUrl`. That is #31. #3 takes an OpenAI-style approach and is much larger. Community options already running: @exfly's server, @gopaljigaur's `decide serve --backends laya`, and @coezbek's hosted endpoint in #32.

**In process, no Python:** @ryuzdev's laya-sdk (weights from HF, browser and Node) and @niutech's pointer to Laya ONNX with Transformers.js.

One caution for the in-process route: the answer is more than the raw logits. Marker positions, the per-bucket temperatures and the `[0.5, 5]` clamp all shape the probabilities. A quick parity check is to run a handful of states through both the Python package and the JS port and compare `probabilities` to about three decimals. If they agree, the port is faithful.

@ryuzdev, if you want it listed in the README's community section, a small PR is welcome.

**Comment by ryuzdev** [comment]
No worries @NandhaKishorM. If you want I can give you maintainer in both repo and `laya-sdk` NPM package since the IP belongs to you and I respect the work 🫡 Contact me on any channel if you will.

**Comment by sirgio03** [comment]
Hey @NandhaKishorM! Thanks for the thoughtful summary and clear roadmap direction !

Just wanted to share a quick clarification regarding PR #3:

Jev Wire Format (POST /v1/systemone): PR #3 actually already implements the exact TypeSafe Jev /v1/systemone wire protocol as a first-class endpoint! Any existing TypeSafe client (@typesafe-ai/sdk, TypeSafe Python SDK, etc.) works out of the box with zero code changes simply by pointing baseUrl to http://localhost:8000.

PR Size & Decoupling js/: The large diff was almost entirely from bundling the TypeScript SDK in the js/ directory (package-lock.json alone was ~2,800 lines). If you prefer keeping the core repository strictly focused on Python, I am more than happy to remove the js/ folder from PR #3 and maintain it as an external community package instead. That brings the PR down to just the lean Python FastAPI server (~600 lines) with non-blocking async inference (run_in_executor) and concurrency protection.

Status: @emiliano-go and I have been collaborating closely over the past few days to refine and test it, and we just synced with the latest v0.3.6 release with all 350+ tests passing (100% green).

I've also updated the PR title and description over on #3 to make the Jev wire compatibility clear. Happy to adapt it however you think best serves Laya! 🙌

**Comment by NandhaKishorM** [comment]
Thanks @sirgio03, and thank you for the collaboration with @emiliano-go.

Where it landed: #31 is now the repo's server and shipped in 0.3.7. The deciding factor was the schema. #31 passes the request straight to `Router.predict`, so it accepts every shape the real Jev API does, and I verified list-shaped criteria and unknown fields against it.

Two parts of your work would be valuable on top of it, if you're up for it:
1. **`run_in_executor` and concurrency protection.** #31 currently calls the router synchronously inside the async endpoint, so your executor approach would be a real improvement as a small PR against `laya/serve.py`.
2. **The TypeScript SDK as a community package** pointed at `/v1/systemone`. I'd be glad to link it in the README's Community Tools section.

For everyone else here, the options today are:
- `laya-serve` with any TypeSafe client;
- @ryuzdev's laya-sdk for in-process use;
- #198, a typed SDK I've asked to target the same endpoint.
